Bluetooth: gatt: notify application of MTU updates

JordanYates opened the pull request:
Add a new callback to `struct bt_conn_cb` to allow applications to track
changes to the ATT MTU. This callback is only called when the remote
side changes the MTU. The application is already notified of a changed
MTU if it initiated the update via the `struct bt_gatt_exchange_params`
callback.

Fixes #32035.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

joerchan commented on include/bluetooth/conn.h:
```diff
@@ -800,6 +800,16 @@ struct bt_conn_cb {
+	 *  @param conn Connection object.
+	 *  @param mtu Updated MTU.
+	 */
+	void (*mtu_updated)(struct bt_conn *conn, uint16_t mtu);
```
I see the value in adding this callback. Especially because this would make it easier when you have multiple GATT profiles and you want something to happen once the MTU has been updated.

This is related to the fact that the MTU exchange should only be initiated once, and multiple profiles might attempt to do just that. At the moment we don't enforce this rule.

 I would therefore like to extend this so that it also called when it has been initiated locally, i.e as a GATT client.
 Also we should make sure it is only called once per connection, that would make it easier for applications, since the ATT mtu should end up as the same in both procedures (as server and client).
 
 Also the name should be updated to reflect that it is the ATT MTU that has changed. MTU of for example L2CAP Channels are handled differently.

JordanYates replied:
> This is related to the fact that the MTU exchange should only be initiated once

Is this part of the specification? I was happily updating the MTU multiple times per connection through the nRF Connect app when testing this PR

JordanYates replied:
> Also the name should be updated to reflect that it is the ATT MTU that has changed

The naming was chosen to be consistent with `bt_conn_get_mtu`. I 100% agree that it is not clear. Is there more value in being consistent or in making the new function clearer?

joerchan replied:
> The naming was chosen to be consistent with bt_conn_get_mtu. I 100% agree that it is not clear. Is there more value in being consistent or in making the new function clearer?

Let's get some more opinions on this before we make this decision.

> Is this part of the specification? I was happily updating the MTU multiple times per connection through the nRF Connect app when testing this PR

Yes, it is part of the specification, but the API does not enforce this rule at the moment.

Vudentz replied:
> > This is related to the fact that the MTU exchange should only be initiated once
> 
> Is this part of the specification? I was happily updating the MTU multiple times per connection through the nRF Connect app when testing this PR

It has a recommendation that MTU exchange is only run once, anyway most of the time having another exchanges makes no different as the values won't change.

joerchan replied:
@Vudentz There is no recommendation, it is a rule.
"This request shall only be sent once during a connection by the client"

jhedberg reviewed (changes requested):
This looks like layer violation to me. bt_conn only abstracts the ACL/SCO/ISO/etc baseband connection. On top of that you could have various protocols, although in practice it's always L2CAP, and on top of that you can have yet another variety of protocols out of which ATT is one possible option.

JordanYates commented:
> This looks like layer violation to me.

True, I didn't consider the layering or alternate protocol implications.
I can see two options for this functionality then.
1. Define a new callback structure in `<gatt.h>`, analogous to `struct bt_conn_cb`, which contains the MTU callback.
2. Add an MTU callback field to `struct bt_gatt_service`, so that services that care are notified of MTU changes.

Commit:
Bluetooth: gatt: callbacks on ATT MTU update

Adds a new callback structure to `<gatt.h>` for receiving notifications
of ATT MTU updates. This callback is called regardless of whether the
MTU update was initiated locally or remotely.

Fixes #32035.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
Implementation has moved from the conn files to gatt files.

Vudentz commented on include/bluetooth/gatt.h:
```diff
@@ -181,6 +181,21 @@ struct bt_gatt_include {
+	 *  @param conn Connection object.
+	 *  @param mtu Updated ATT MTU.
+	 */
+	void (*att_mtu_updated)(struct bt_conn *conn, uint16_t mtu);
```
This won't work when there are EATT channels stabilished, since there you can have different MTU for tx/rx and also different sizes between each channe, or you intend to pass only the highest of the channels?

JordanYates replied:
I wasn't aware of EATT so there was no intention at all. Are ATT and EATT separate enough that it would make sense to have
```
    void (*att_mtu_updated)(struct bt_conn *conn, uint16_t mtu);
    void (*eatt_mtu_updated)(struct bt_conn *conn, uint8_t channel, uint16_t tx_mtu, uint16_t rx_mtu);
```
Or is there some special eatt reference to signify the old att setup that could be used? Channel = 0 or something?

joerchan replied:
@JordanYates  Seeing as we only have `bt_att_get_mtu` and it returns the highest MTU of all the channel it seems most correct to me to only have one callback, and call it each time the highest possible MTU has increased.

The EATT channels are L2CAP channels and they don't have the MTU exchange procedure, instead the MTU is set based on the L2CAP MTU on the channel.
So in order to call the callback callback for this we would have to use the L2CAP connected callback on the ATT channel.

Alternatively we could document that this is only called on the Fixed ATT channel (or Unenhanced) as a simplification.

carlescufi replied:
> @JordanYates Seeing as we only have `bt_att_get_mtu` and it returns the highest MTU of all the channel it seems most correct to me to only have one callback, and call it each time the highest possible MTU has increased.

Agreed with this, @JordanYates could you please implement it this way?

JordanYates replied:
> Agreed with this, @JordanYates could you please implement it this way?

I believe the updates implement what you want.

To pick up the EATT channels code is run in `bt_att_connected`, which in turn means the callback is run (due to the fixed ATT channel) before the standard `bt_conn` `connected` callback is run. If this is not desired I  can move that code into an EATT conditional.

The maximum MTU detection code doesn't detect when the MTU has dropped. My understanding is that this could only happen if an EATT channel is closed without the connection dropping. Is this possible?

Vudentz commented on subsys/bluetooth/host/att.c:
```diff
@@ -558,6 +558,9 @@ static uint8_t att_mtu_req(struct bt_att_chan *chan, struct net_buf *buf)
 
 	BT_DBG("Negotiated MTU %u", chan->chan.rx.mtu);
+
+	bt_gatt_att_mtu_updated(conn, chan->chan.rx.mtu);
```
EATT channels don't use MTU exchange to change its MTU.

carlescufi commented:
@JordanYates will you follow this up after the review?

JordanYates commented:
> @JordanYates will you follow this up after the review?

Sure. I have no knowledge of EATT thought so I asked @Vudentz for some guidance.
I was waiting on that before implementing something that would probably be wrong.

Commit:
samples: Bluetooth: peripheral: add MTU callback

Add a demonstration of the updated ATT MTU callback to the Bluetooth
peripheral sample.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Vudentz commented on include/bluetooth/gatt.h:
```diff
@@ -181,6 +181,22 @@ struct bt_gatt_include {
+	 *  @param tx Updated TX ATT MTU.
+	 *  @param rx Updated RX ATT MTU.
+	 */
+	void (*att_mtu_updated)(struct bt_conn *conn, uint16_t tx, uint16_t rx);
```
How does this fixes any problem that the likes of bt_gatt_get_mtu wouldn't? If there is a race with the MTU change you would likely have this values outdated on the application side.

JordanYates replied:
When do you call `bt_gatt_get_mtu` then? You have no way of knowing when the value is likely to change.

Can you expand on the MTU race condition? In my mind this is inherently much safer because instead of polling `bt_gatt_get_mtu`, where the value could change 1 cycle after the call, you are only notified after the change. If the MTU only changes once on a connection like it should, I'm not sure how thats racey.

jhedberg commented on include/bluetooth/gatt.h:
```diff
@@ -181,6 +181,22 @@ struct bt_gatt_include {
+	 */
+	void (*att_mtu_updated)(struct bt_conn *conn, uint16_t tx, uint16_t rx);
+
+	struct bt_gatt_cb *_next;
```
Even though this is fairly trivial, I think we try to avoid custom linked list implementations and use `sys_snode_t` instead.

jhedberg reviewed (changes requested):
Looks generally alright to me, except for the custom linked list implementation which should be replaced with Zephyr's own singly-linked list solution.

JordanYates commented:
@jhedberg custom list replaced with `sys_slist_t`
